src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Gemm.cpp - Gemm Op --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX Gemm operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "llvm/ADT/SmallVector.h"
+#include <cstdint>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+
+
+class ONNXGemmOpLoweringToTOSA : public OpConversionPattern<ONNXGemmOp> {
+public:
+  using OpConversionPattern<ONNXGemmOp>::OpConversionPattern;
+  LogicalResult rewriteToTosaFC(ONNXGemmOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const {
+    Value A = adaptor.A();
+    Value B = adaptor.B();
+    Value C = adaptor.C();
+
+    Type resultType = getTypeConverter()->convertType(op.getResult().getType());
+
+    // If no bias is given to the GEMM operator, we create a 1D bias with all
+    // zeroes
+    if (C.getType().isa<mlir::NoneType>()) {
+      // B is supposed to have 
+      ArrayRef<int64_t> cformat(B.getType().cast<TensorType>().getShape()[1]);
+      std::vector<float> elements = {};
+      for (int i = 0; i < cformat[0]; ++i)
+        elements.push_back(0.0F);
+      C = mlir::tosa::getConstTensor<float>(rewriter, op, elements, cformat).value();
+    }
+
+    rewriter.replaceOpWithNewOp<tosa::FullyConnectedOp>(op, resultType, A, B, C);
+    return success();
+  }
+
+  /// The GEMM can be described as a FullyConnected operator
+  /// Y = AB^T + C if we perform a transpose on B only with
+  /// alpha and beta factors set to 1.
+  /// The constraints on rank were taken from the torch implementation check
```

Comment:
This comment won't make much sense when upstreaming, as we don't plan to upstream our onnx-to-torch implementation.